lib/FreeRTOS/freertos-plus-mqtt/agent_command_pool.c

```diff
@@ -0,0 +1,165 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://www.FreeRTOS.org
+ * https://aws.amazon.com/freertos
+ *
+ */
+
+/**
+ * @file agent_command_pool.c
+ * @brief Implements functions to obtain and release commands.
+ */
+
+/* Standard includes. */
+#include <string.h>
+#include <stdio.h>
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "semphr.h"
+
+/* Header include. */
+#include "agent_command_pool.h"
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief The pool of command structures used to hold information on commands (such
+ * as PUBLISH or SUBSCRIBE) between the command being created by an API call and
+ * completion of the command by the execution of the command's callback.
+ */
+static Command_t commandStructurePool[ MQTT_COMMAND_CONTEXTS_POOL_SIZE ];
+
+/**
+ * @brief A counting semaphore used to guard the pool of Command_t structures.  To
+ * obtain a structure first decrement the semaphore count.  To return a structure
+ * increment the semaphore count after the structure is back in the pool.
+ */
+static SemaphoreHandle_t freeCommandStructMutex = NULL;
+
+static volatile bool initializedSemaphore = false;
+
+/*-----------------------------------------------------------*/
+
+static void initializePool()
+{
+    bool destroySemaphore = true;
+    SemaphoreHandle_t tempMutex = xSemaphoreCreateCounting( MQTT_COMMAND_CONTEXTS_POOL_SIZE, MQTT_COMMAND_CONTEXTS_POOL_SIZE );
+
+    taskENTER_CRITICAL();
+    {
+        if( !initializedSemaphore )
+        {
+            memset( ( void * ) commandStructurePool, 0x00, sizeof( commandStructurePool ) );
+            freeCommandStructMutex = tempMutex;
+            destroySemaphore = false;
+        }
+
+        initializedSemaphore = true;
+    }
+    taskEXIT_CRITICAL();
+
+    /* Destroy the semaphore if it was not used. */
+    if( destroySemaphore )
+    {
+        vSemaphoreDelete( tempMutex );
+    }
+}
```

Comment:
This bugs me.  To my mind the critical section should only encompass toggling the flag.  There is no sin in doing an outer check on the flag, and then an inner one protected by the critical section.  In which case there would be no need for the destroySemaphore business.

Reply:
Could you clarify what you mean by this? I don't think it would be like:
```
if( !initialized )
{
    taskENTER_CRITICAL();
    {
        if( !initializedSemaphore )
        {
             initializedSemaphore = true;
        }
    }
    taskEXIT_CRITICAL();
}
```
where do we create the semaphore?

Reply:
something like:

```c
typedef enum
{
    notDone,
    pending,
    done
} initStatus_t;

initStatus_t initialized = notDone;

initStatus_t doInit()
{
    bool owner = false;

    if( initialized == notDone )
    {
        /* use shortest possible critical section */
        taskENTER_CRITICAL();
        if( initialized == notDone )
        {
            initialized = pending;
            owner = true;
        }
        taskEXIT_CRITICAL();
    }

    if( owner == true )
    {
        /* do all creation and setup; no critical section needed */
        initialized = done;
    }

    return initialized;
}
```

Reply:
Updated